src/lib/Libifl/pbsD_connect.c

```diff
@@ -470,13 +335,226 @@ __pbs_connect_extend(char *server, char *extend_data)
+ * @param[in,out]	conn_arr - array of svr_conn_t
+ * @param[in]		extend_data - any additional data relevant for connection
+ *
+ * @return	int
+ * @retval	-1 for error
+ * @retval	fd of connection
+ */
+static int
+connect_to_server(int idx, svr_conn_t *conn_arr, char *extend_data)
+{
+	int sd = conn_arr[idx].sd;
+	struct sockaddr_in my_sockaddr;
+
+	/* bind to pbs_public_host_name if given  */
+	if (pbs_conf.pbs_public_host_name) {
+		if (get_hostsockaddr(pbs_conf.pbs_public_host_name, &my_sockaddr) != 0)
+			return -1; /* pbs_errno was set */
+		/* my address will be in my_sockaddr,  bind the socket to it */
+		my_sockaddr.sin_port = 0;
+		if (bind(sd, (struct sockaddr *)&my_sockaddr, sizeof(my_sockaddr)) != 0) {
+			return -1;
+		}
+	}
+
+	if (conn_arr[idx].state != SVR_CONN_STATE_UP) {
+		if ((sd = tcp_connect(conn_arr[idx].name, conn_arr[idx].port, extend_data)) != -1) {
+			conn_arr[idx].state = SVR_CONN_STATE_UP;
+			conn_arr[idx].sd = sd;
+		} else
+			conn_arr[idx].state = SVR_CONN_STATE_DOWN;
+	}
+
+	return sd;
+}
+
+/**
+ * @brief	To connect to all the servers
+ *
+ * @param[in]	svrhost - valid host name of one of the servers
+ * @param[in]	port - port of the server to connect to (considered if server_name is not NULL)
+ * @param[in]	extend_data
+ *
+ * @return int
+ * @retval >0 - success
+ * @retval -1 - error
+ */
+static int
+connect_to_servers(char *svrhost, uint port, char *extend_data)
+{
+	int i;
+	int fd = -1;
+	int ret_fd = -1;
+	int num_conf_servers = get_num_servers();
+	svr_conn_t *svr_connections = get_conn_servers(-1);
+
+	if (svr_connections == NULL)
+		return -1;
+
+	if (svrhost != NULL) {
+		struct sockaddr_in tmp_sockaddr;
+
+		/* Check that this host is reachable */
+		if (get_hostsockaddr(svrhost, &tmp_sockaddr) != 0) {
+			pbs_errno = PBSE_BADHOST;
+			return -1;
+		}
+	}
+
+	/* Try to connect to all servers in the cluster */
+	for (i = 0; i < num_conf_servers; i++) {
+		fd = connect_to_server(i, svr_connections, extend_data);
+
+		if (ret_fd == -1 && fd != -1)
+			ret_fd = fd;
+	}
+
+	return ret_fd;
+}
+
+/**
+ * @brief	Makes a PBS_BATCH_Connect request to 'server'.
+ *
+ * @param[in]   server - the hostname of the pbs server to connect to.
+ * @param[in]   extend_data - a string to send as "extend" data.
+ *
+ * @return int
+ * @retval >= 0	index to the internal connection table representing the
+ *		connection made.
+ * @retval -1	error encountered setting up the connection.
+ */
+int
+__pbs_connect_extend(char *server, char *extend_data)
+{
+	char server_name[PBS_MAXSERVERNAME + 1];
+	unsigned int server_port;
+
+	/* initialize the thread context data, if not already initialized */
+	if (pbs_client_thread_init_thread_context() != 0)
+		return -1;
+
+	if (pbs_loadconf(0) == 0)
+		return -1;
+
+	server = PBS_get_server(server, server_name, &server_port);
+	if (server == NULL) {
+		pbs_errno = PBSE_NOSERVER;
+		return -1;
+	}
+
+	return connect_to_servers(server_name, server_port, extend_data);
```

Comment:
Not sure but shouldn't we need to do this for pbs_connect_noblk() also?

Reply:
Thanks for mentioning that, I hadn't thought about it. AFAIK, right now we don't have plans to support a multi-multi-server configuration, so when the scheduler calls pbs_connect_noblk(), it will connect to the one remote server that it wants to connect to, and then can call pbs_disconnect to disconnect from that single server. I have made a small change to pbs_disconnect to support this. Other than that, we don't need to make any changes to pbs_connect_noblk() to support multi-server since it's an internal interface. Let me know what you think.